website/content/docs/agent/telemetry.mdx

```diff
@@ -269,6 +269,57 @@ resources will still work.
+
+In Consul each write operation will turn into a single Raft log to be committed. Raft will process these
+logs and store them within Bolt DB in batches. Each call to store logs within Bolt DB is measured to record how long
+it took as well as how many logs were contained in the batch. Writing logs is this fashion is serialized so that
+a subsequent log storage operation can only be started after the previous one completed. Therefore the maximum number
+of log storage operations that can be performed each second can be calculated with the following equation: 
+`1000 ms / consul.raft.boltdb.storeLogs ms/op`. From there we can extrapolate the maximum number of Consul writes
+per second by multiplying that value by the `consul.raft.boltdb.logsPerBatch` metric's value. When log storage 
+operations are becoming slower you may not see an immediate decrease in write throughput to Consul due to increased 
+batch sizes of the each operation. However, the max batch size allowed is 64 logs. Therefore if the `logsPerBatch`
+metric is near 64 and the `storeLogs` metric is seeing increased time to write each batch to disk, then it is likely 
+that increased write latencies and other errors may occur.
+
+There can be a number of potential issues that can cause this. Often times it could be performance of the underlying
+disks that is the issue. Other times it may be caused by Bolt DB behavior. Bolt DB keeps track of free space within
+the raft.db file. When needing to allocate data it will use existing free space first before further expanding the
+file. By default, Bolt DB will write a data structure containing metadata about free pages within the DB to disk for
+every log storage operation. Therefore if the free space within the database grows excessively large, such as after
+a large spike in writes beyond the normal steady state and a subsequent slow down in the write rate, then Bolt DB
+could end up writing a large amount of extra data to disk for each log storage operation. This has the potential
+to drastically increase disk write throughput, potentially beyond what the underlying disks can keep up with. To
+detect this situation you can look at the `consul.raft.boltdb.freelistBytes` metric. This metric is a count of
+the extra bytes that are being written for each log storage operation beyond the log data itself. While not a clear
+indicator of an actual issue, thiss metric can be used to diagnose why the `consul.raft.boltdb.storeLogs` metric
+is high. 
+
+If Bolt DB log storage performance becomes an issue and is caused by free list management then setting
+[`raft_boltdb.NoFreelistSync`](/docs/agent/options#NoFreelistSync) to `true` in the servers configuration
+may help to reduce disk IO and log storage operation times. Disabling free list syncing will however increase
+the startup time for a server as it must scan the raft.db file for free space instead of loading the already
+populated free list structure.
```

Comment:
Do we have a vague sense for what the time tradeoff here is between the _every op_ vs _startup_ side of things?

Reply:
In my rough testing with a raft.db with several GB of free space and ~500 MB of real data the time difference was negligible. It added milliseconds. I have heard things of there being some cases where this might not be the case but I _think_ they might only be when the bolt db database gets enormous. Our usage of boltdb as a wal and frequent truncation coupled with our architecture basically mandating reasonably fast ssds for storage should mean that the file can never grow to such a size where it would be problematic. 

Additionally the free pages should generally/always be contiguous due to our utilization of boltdb as a dynamically sized ring buffer of logs. I suspect this might reduce the time to recreate the freelist vs something like Vaults usage of bolt db where they do more random access and where the DB grows a lot larger.